### DISCUSSION TITLE
What is the equivalent of IO monad?

### DISCUSSION THREAD

**Discussion by muratkeremozcan** [bloodyowl/boxed#43] [Q&A] [answered]
I am learning more about Boxed, love it so far.

Option is like Maybe
Result is like Either

What would be the equivalent of IO monad? Would you shed some light?

**Comment by bloodyowl** [comment, marked as answer]
Hi!

We don't have an equivalent in Boxed. Our goal isn't to mimic all the common monads from FP languages, but only the minimal set that makes the most practical sense in a JS ecosystem 😊

**Reply by muratkeremozcan** [reply]
Understood. 

For example we have an error handling example (From James Sinclair's new book) https://github.com/muratkeremozcan/books/blob/master/99_MuratsNinjaExercises/fp-skeptics-guide/03-monad.js

Reproducing it with swan is fine https://github.com/muratkeremozcan/books/blob/master/99_MuratsNinjaExercises/fp-skeptics-guide/03-monad.ts until we get to some other common monad features like  `ap`, `orElse`, `lift` . What would be your recommendation in this case? Should we use `fp-ts` together with swan where swan chooses to be minimal?

**Reply by bloodyowl** [reply]
What you wanna do looks something like this:

```ts
const getUserBanner = (user?: User) => {
  return Option.fromNullable(user)
    .flatMap(user => Option.fromNullable(user.accountDetails))
    .flatMap(accountDetails => Option.fromNullable(accountDetails.address))
    .flatMap(address => Option.fromNullable(address.province))
    .flatMap(province => Option.fromNullable(banners[province]))
}

const customizeBanner = (user?: User) => {
  const element: Option<HTMLImageElement> = Option.fromNullable(document.querySelector(".banner > img"))
  const banner = getUserBanner(user)
  Option.allFromDict({element, banner})
    .match({
      Some: ({element, banner}) => {
        element.src = banner;
      },
      None: () => {
        // do nothing
      }
    })
}
```
